File: plugins/skill-intake/scripts/validate_mermaid.js
#!/usr/bin/env node
'use strict';

const fs = require('node:fs');
const path = require('node:path');

const HEADERS = ['flowchart', 'graph', 'sequenceDiagram', 'classDiagram', 'stateDiagram', 'stateDiagram-v2',
  'erDiagram', 'gantt', 'pie', 'journey', 'mindmap', 'timeline', 'quadrantChart'];

function jaLen(s) {
  let n = 0;
  for (const ch of s) n += ch.charCodeAt(0) > 127 ? 1 : 0.5;
  return n;
}

function extractNodeLabels(src) {
  const labels = [];
  const reBracket = /\[([^\]\n]+)\]/g;
  const reParen = /\(([^)\n]+)\)/g;
  let m;
  while ((m = reBracket.exec(src))) labels.push(m[1]);
  while ((m = reParen.exec(src))) labels.push(m[1]);
  return labels;
}

function validate(src) {
  const errors = [];
  const warnings = [];
  const trimmed = src.trim();
  if (!trimmed) return { ok: false, errors: ['empty diagram'], warnings, node_count: 0 };
  const firstWord = trimmed.split(/\s+/)[0];
  if (!HEADERS.some((h) => firstWord.startsWith(h))) {
    errors.push(`unknown diagram header: ${firstWord}`);
  }
  const lines = trimmed.split(/\n/).filter((l) => l.trim() && !l.trim().startsWith('%%'));
  const bodyLines = lines.slice(1);
  const nodeCount = bodyLines.length;
  if (nodeCount < 5 || nodeCount > 9) {
    warnings.push(`node/line count ${nodeCount} outside 7±2`);
  }
  const labels = extractNodeLabels(trimmed);
  for (const lbl of labels) {
    if (jaLen(lbl) > 10) warnings.push(`label too long: "${lbl}"`);
  }
  return { ok: errors.length === 0, errors, warnings, node_count: nodeCount, label_count: labels.length };
}

function main(argv) {
  const file = argv[2];
  if (!file) { process.stderr.write('usage: validate_mermaid.js <diagram.mmd>\n'); return 2; }
  let src;
  try { src = fs.readFileSync(path.resolve(file), 'utf8'); }
  catch (e) { process.stderr.write(`input error: ${e.message}\n`); return 2; }
  const r = validate(src);
  process.stdout.write(JSON.stringify(r, null, 2) + '\n');
  return r.ok ? 0 : 1;
}

module.exports = { validate, jaLen };

if (require.main === module) {
  process.exit(main(process.argv));
}

File: plugins/skill-intake/scripts/validate_mermaid.py
#!/usr/bin/env python3
"""Lightweight Mermaid syntax sanity validator (no external CLI)."""
from __future__ import annotations

import json
import re
import sys
from pathlib import Path
from typing import Any

HEADERS = [
    'flowchart', 'graph', 'sequenceDiagram', 'classDiagram', 'stateDiagram', 'stateDiagram-v2',
    'erDiagram', 'gantt', 'pie', 'journey', 'mindmap', 'timeline', 'quadrantChart',
]

_BRACKET = re.compile(r'\[([^\]\n]+)\]')
_PAREN = re.compile(r'\(([^)\n]+)\)')


def ja_len(s: str) -> float:
    n = 0.0
    for ch in s:
        n += 1 if ord(ch) > 127 else 0.5
    return n


def extract_node_labels(src: str) -> list[str]:
    return _BRACKET.findall(src) + _PAREN.findall(src)


def validate(src: str) -> dict[str, Any]:
    errors: list[str] = []
    warnings: list[str] = []
    trimmed = src.strip()
    if not trimmed:
        return {'ok': False, 'errors': ['empty diagram'], 'warnings': warnings, 'node_count': 0}
    first_word = trimmed.split()[0]
    if not any(first_word.startswith(h) for h in HEADERS):
        errors.append(f'unknown diagram header: {first_word}')
    lines = [l for l in trimmed.split('\n') if l.strip() and not l.strip().startswith('%%')]
    body_lines = lines[1:]
    node_count = len(body_lines)
    if node_count < 5 or node_count > 9:
        warnings.append(f'node/line count {node_count} outside 7±2')
    labels = extract_node_labels(trimmed)
    for lbl in labels:
        if ja_len(lbl) > 10:
            warnings.append(f'label too long: "{lbl}"')
    return {
        'ok': len(errors) == 0,
        'errors': errors,
        'warnings': warnings,
        'node_count': node_count,
        'label_count': len(labels),
    }


def main(argv: list[str]) -> int:
    if len(argv) < 1:
        sys.stderr.write('usage: validate_mermaid.py <diagram.mmd>\n')
        return 2
    try:
        src = Path(argv[0]).resolve().read_text(encoding='utf-8')
    except Exception as e:
        sys.stderr.write(f'input error: {e}\n')
        return 2
    r = validate(src)
    sys.stdout.write(json.dumps(r, ensure_ascii=False, indent=2) + '\n')
    return 0 if r['ok'] else 1


if __name__ == '__main__':
    sys.exit(main(sys.argv[1:]))
